pkg/utils/overriding/merging_test.go

```diff
@@ -0,0 +1,372 @@
+				},
+				Attributes: attributesPkg.Attributes{}.FromMap(map[string]interface{}{
+					"parent": true,
+					"plugin": true,
+					"main":   true,
+				}, nil),
+				Commands: []dw.Command{
+					{
+						Id: "parentCommand",
+						CommandUnion: dw.CommandUnion{
+							Exec: &dw.ExecCommand{
+								WorkingDir: "dir",
+							},
+						},
+					},
+					{
+						Id: "pluginCommand",
+						CommandUnion: dw.CommandUnion{
+							Exec: &dw.ExecCommand{
+								WorkingDir: "dir",
+							},
+						},
+					},
+					{
+						Id: "mainCommand",
+						CommandUnion: dw.CommandUnion{
+							Exec: &dw.ExecCommand{
+								WorkingDir: "dir",
+							},
+						},
+					},
+				},
+				Components: []dw.Component{
+					{
+						Name: "parentComponent",
+						ComponentUnion: dw.ComponentUnion{
+							Container: &dw.ContainerComponent{
+								Container: dw.Container{
+									Image: "image",
+								},
+							},
+						},
+					},
+					{
+						Name: "pluginComponent",
+						ComponentUnion: dw.ComponentUnion{
+							Container: &dw.ContainerComponent{
+								Container: dw.Container{
+									Image: "image",
+								},
+							},
+						},
+					},
+					{
+						Name: "mainComponent",
+						ComponentUnion: dw.ComponentUnion{
+							Container: &dw.ContainerComponent{
+								Container: dw.Container{
+									Image: "image",
+								},
+							},
+						},
+					},
+				},
+				Events: &dw.Events{
+					DevWorkspaceEvents: dw.DevWorkspaceEvents{
+						PreStart:  []string{},
+						PostStart: []string{"post-start-parent"},
+						PreStop:   []string{},
+						PostStop:  []string{"post-stop-main", "post-stop-parent", "post-stop-plugin"},
+					},
+				},
+			},
+			wantErr: nil,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			mergedContent, err := MergeDevWorkspaceTemplateSpec(tt.mainContent, tt.parentFlattenedContent, tt.pluginFlattenedContents...)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+
+			assert.Equal(t, tt.expected, mergedContent, "The two values should be the same.")
+		})
+	}
+}
+
+func mergingPatchTest(main, parent, expected []byte, expectedError string, plugins ...[]byte) func(t *testing.T) {
+	return func(t *testing.T) {
+		result, err := MergeDevWorkspaceTemplateSpecBytes(main, parent, plugins...)
+		if err != nil {
+			compareErrorMessages(t, expectedError, err.Error(), "wrong error")
+			return
+		}
+		if expectedError != "" {
+			t.Error("Expected error but did not get one")
+			return
+		}
+
+		resultJson, err := json.Marshal(result)
+		if err != nil {
+			t.Error(err)
+		}
+		resultYaml, err := yaml.JSONToYAML(resultJson)
+		if err != nil {
+			t.Error(err)
+		}
+
+		expectedJson, err := yamlMachinery.ToJSON(expected)
+		if err != nil {
+			t.Error(err)
+		}
+		expectedYaml, err := yaml.JSONToYAML(expectedJson)
+		if err != nil {
+			t.Error(err)
+		}
+
+		assert.Equal(t, string(expectedYaml), string(resultYaml), "The two values should be the same.")
```

Comment:
this were written prev, i just cut-paste them to a new file, but i'll update it

Reply:
I think the original intention of the prev author was to have the result.yaml file contents sorted when comaring, the yaml file bytes is converted to JSON and then re-converted to YAML to maintain the order

```golang
                expectedJson, err := yamlMachinery.ToJSON(expected)
		if err != nil {
			t.Error(err)
		}
		expectedYaml, err := yaml.JSONToYAML(expectedJson)
		if err != nil {
			t.Error(err)
		}

		assert.Equal(t, string(expectedYaml), string(resultYaml), "The two values should be the same.")
``` 

I noticed that it results in different order otherwise:

```
FAIL: TestMerging/no-parent (0.00s)
        /Users/maysun/dev/redhat/api/pkg/utils/overriding/merging_test.go:40: 
            	Error Trace:	merging_test.go:40
            	Error:      	Not equal: 
            	            	expected: "components:\n  - name: test-plugin-component\n    container:\n      image: test-plugin-image\n  - name: test-component\n    container:\n      image: test-image\n"
            	            	actual  : "components:\n- container:\n    image: test-plugin-image\n  name: test-plugin-component\n- container:\n    image: test-image\n  name: test-component\n"
            	            	
            	            	Diff:
            	            	--- Expected
            	            	+++ Actual
            	            	@@ -1,8 +1,8 @@
            	            	 components:
            	            	-  - name: test-plugin-component
            	            	-    container:
            	            	-      image: test-plugin-image
            	            	-  - name: test-component
            	            	-    container:
            	            	-      image: test-image
            	            	+- container:
            	            	+    image: test-plugin-image
            	            	+  name: test-plugin-component
            	            	+- container:
            	            	+    image: test-image
            	            	+  name: test-component
            	            	 
            	Test:       	TestMerging/no-parent
            	Messages:   	The two values should be the same.
 ```

Reply:
Huh, strange. I'm fine with keeping it as is but that should definitely have a doc explaining why it's necessary (otherwise someone like me might remove it when I see it in a year ;))

Another option is to compare structs directly, something along the lines of
```go
var expectedTemplate *dw.DevWorkspaceTemplateSpec
err = yaml.Unmarshal(expected, expectedTemplate)
if err != nil {
	t.Error(err)
}
assert.Equal(t, expectedTemplate, result, "The two values should be the same")
```
this makes some sense, since we're already serializing the result -- why not deserialize the expected instead?

If we want to keep the nice diff showing the error, this can be done using `go-cmp` as well -- replace the `assert.Equal` with
```go
assert.True(t, cmp.Equal(expectedTemplate, result), fmt.Sprintf("Output doesn't match expected\n%s", cmp.Diff(expectedTemplate, result)))
```
(I don't remember if `assert.Equal` prints nice diffs by default)

Reply:
Thats a good idea, I've just done what you've suggested and used the template struct to compare. Updated for both merging and overriding tests